extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/HiveFileIndex.scala

```diff
@@ -0,0 +1,269 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.spark.connector.hive
+
+import java.net.URI
+
+import scala.collection.mutable
+
+import org.apache.hadoop.fs.{FileStatus, Path}
+import org.apache.hadoop.hive.ql.metadata.{Partition => HivePartition, Table}
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.{expressions, InternalRow}
+import org.apache.spark.sql.catalyst.catalog.{CatalogTable, CatalogTablePartition, ExternalCatalogUtils}
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, BoundReference, Expression, Predicate}
+import org.apache.spark.sql.connector.catalog.CatalogPlugin
+import org.apache.spark.sql.execution.datasources._
+import org.apache.spark.sql.hive.kyuubi.connector.HiveConnectorHelper.hiveClientImpl
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.StructType
+
+class HiveCatalogFileIndex(
+    sparkSession: SparkSession,
+    val catalogTable: CatalogTable,
+    hiveCatalog: HiveTableCatalog,
+    override val sizeInBytes: Long)
+  extends PartitioningAwareFileIndex(
+    sparkSession,
+    catalogTable.storage.properties,
+    Some(catalogTable.schema)) {
+
+  private val table = catalogTable
+
+  private val partPathToBindHivePart: mutable.Map[PartitionPath, HivePartition] = mutable.Map()
+
+  private val fileStatusCache = FileStatusCache.getOrCreate(sparkSession)
+
+  private lazy val hiveTable: Table = hiveClientImpl.toHiveTable(table)
+
+  private val baseLocation: Option[URI] = table.storage.locationUri
+
+  override def partitionSchema: StructType = table.partitionSchema
+
+  private[hive] def listHiveFiles(partitionFilters: Seq[Expression], dataFilters: Seq[Expression])
+      : (Seq[PartitionDirectory], Map[PartitionDirectory, HivePartition]) = {
+    val fileIndex = filterPartitions(partitionFilters)
+    val partDirs = fileIndex.listFiles(partitionFilters, dataFilters)
+    val partDirToHivePart = fileIndex.partDirToBindHivePartMap()
+    (partDirs, partDirToHivePart)
+  }
+
+  def filterPartitions(filters: Seq[Expression]): HiveInMemoryFileIndex = {
+    if (table.partitionColumnNames.nonEmpty) {
+      val startTime = System.nanoTime()
+      val selectedPartitions: Seq[BindPartition] = {
+        val partitions =
+          listPartitionsByFilter(sparkSession.sessionState.conf, hiveCatalog, table, filters)
```

Comment:
IIUC, this can be simply used by `ExternalCatalogUtils.listPartitionsByFilter(sparkSession.sessionState.conf, hiveCatalog.catalog, table, filters)`